### ISSUE TITLE
importhook without using typechecked

### ISSUE THREAD

**Issue by albertz** [agronholm/typeguard#281]
**Is your feature request related to a problem? Please describe.**
The importhook adds `@typechecked` annotations. This will always add another extra frame to the call stack for every type-checked function.

**Describe the solution you'd like**
I think having such extra frame can be avoided in an importhook.

**Describe alternatives you've considered**
Instead of adding the `@typechecked`, we can directly modify the code of any function to perform the type check. This is still fairly straightforward:
- Right in the beginning of the function, do the checks on all input args. E.g. add some code like:
  ```python
  memo = CallMemo(...)
  check_argument_types(memo)
  ```
  Make sure that `memo` is a unique name.
- For every `return`, add a check.
  * For cases like `return y`, add:
    ```python
    check_return_type(y, memo)
    ```
  * For cases like `return a + b`, we need to add a temporary variable:
    ```python
     retval = a + b
     check_return_type(retval, memo)
     ```
     Make sure that `retval` is a unique name.
- At the very end of the function (maybe check if it is reachable if you want), add:
  ```python
  check_return_type(None, memo)
  ```

**Comment by agronholm** [comment]
Yeah, that extra frame is annoying. There are some cases where it's pretty much unavoidable though, like nested functions, since the function object cannot be resolved from within the function body. But I think these can be detected and explicitly handled with `@typechecked`.

So yeah, I'm on board with this.

**Comment by albertz** [comment]
For some own projects, I use [this logic](https://github.com/albertz/py_better_exchook/blob/bd0325b3ceaa1e4baf5bc832569d926e4fa1e57f/better_exchook.py#L1467) to get back the function object from any stack frame. But not sure if you want to do that. It's a bit ugly, slow, depends on behavior of CPython.

In nested functions, if the nested function has a name, you probably still can access it, or not?
```python
def func():
  def nested_func():
    memo = CallMemo(nested_func, ...)
    ...
```

**Comment by agronholm** [comment]
> For some own projects, I use [this logic](https://github.com/albertz/py_better_exchook/blob/bd0325b3ceaa1e4baf5bc832569d926e4fa1e57f/better_exchook.py#L1467) to get back the function object from any stack frame. But not sure if you want to do that. It's a bit ugly, slow, depends on behavior of CPython.

This looks like you're trying to find the first function object matching the code. But that doesn't work reliably since you could have multiple function objects reachable from the GC, and you have no way to tell which was just called. Worse, the function object could be gone altogether.

> In nested functions, if the nested function has a name, you probably still can access it, or not?

No, you can't. The function object could simply be **gone**, not available from the garbage collector or anywhere else.

**Comment by albertz** [comment]
Right, my function is definitely not 100% reliable, although it still works almost always in the usual cases. For my use cases, like printing debug info to the user, this is maybe acceptable. But I can understand we would not want this here.

> > In nested functions, if the nested function has a name, you probably still can access it, or not?
> 
> No, you can't. The function object could simply be **gone**, not available from the garbage collector or anywhere else.

No, this is wrong. In my example, as I access `nested_func` in the code (`memo = CallMemo(nested_func, ...)`), it will be in the function closure, so it is always preserved. But you need to make sure that it is being referred to, by adding such code (`memo = ...`). But this is possible in the importhook.

**Comment by agronholm** [comment]
> No, this is wrong. In my example, as I access `nested_func` in the code (`memo = CallMemo(nested_func, ...)`), it will be in the function closure, so it is always preserved. But you need to make sure that it is being referred to, by adding such code (`memo = ...`). But this is possible in the importhook.

Ok, you may be right. Indeed this is only possible with the import hook. I'll try to get this done this weekend and get back to you with the results.

**Comment by agronholm** [comment]
I'm making some progress. Argument checking and return type checking are working in module level functions and nested functions, but not yet in class/instance methods.

**Comment by agronholm** [comment]
Class/instance methods are now working too, but not in nested classes. I'll call it a day.

**Comment by albertz** [comment]
Very nice. I'm looking forward to it.

**Comment by agronholm** [comment]
Nested classes, generators and async generators work now.

**Comment by albertz** [comment]
Very nice!

As I see, you use the hardcoded name `_call_memo`. I guess it is unlikely that any other code has used it. However, if there is any code using that name, that would break it, right? I think you can actually check for all used names inside the module (both globals and locals) and then safely use really a unique name. Or even just search for the string occurance in the module source code.

Also, I think you miss one final `check_return_type` at the very end of a function, right? E.g. in a function without a `return`, or with only optional return, like:
```python

def func(a: bool, b: int) -> Optional[int]:
    if a:
        return b
```

**Comment by agronholm** [comment]
Both issues fixed in 20c0a135 and 395f2124, respectively.

**Comment by albertz** [comment]
Very nice. Thanks a lot.
